src/middleware/compression.rs

```diff
@@ -0,0 +1,233 @@
+#![cfg(feature = "compression")]
+use brotli2::read::{BrotliDecoder, BrotliEncoder};
+use bytes::{Buf, Bytes, IntoBuf};
+use flate2::read::{DeflateDecoder, DeflateEncoder, GzDecoder, GzEncoder};
+use futures::future::FutureObj;
+use futures::stream::StreamExt;
+use http::{
+    header::{HeaderValue, ACCEPT_ENCODING, CONTENT_ENCODING},
+    status::StatusCode,
+    HeaderMap,
+};
+use http_service::Body;
+use lazy_static::lazy_static;
+use regex::Regex;
+use std::io::Read;
+
+use crate::{
+    middleware::{Middleware, Next},
+    response::IntoResponse,
+    Context, Error, Response,
+};
+
+lazy_static! {
+    static ref RE_ACCEPT_ENCODING: Regex =
+        Regex::new(r"(?P<a>[\w\*]+)(;q=(?P<q>\d+(\.\d{1,2})?))?(\s*,\s*)?").unwrap();
+}
+
+/// Encodings for Compression/Decompression middleware.
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum Encoding {
+    Gzip,
+    Deflate,
+    Identity,
+    Br,
+}
+
+impl Encoding {
+    fn header_value(self) -> HeaderValue {
+        match self {
+            Encoding::Gzip => HeaderValue::from_str("gzip").unwrap(),
+            Encoding::Deflate => HeaderValue::from_str("deflate").unwrap(),
+            Encoding::Identity => HeaderValue::from_str("identity").unwrap(),
+            Encoding::Br => HeaderValue::from_str("br").unwrap(),
+        }
+    }
+}
+#[derive(Default)]
+pub struct EncodeSettings {
+    gzip_level: flate2::Compression,
+    deflate_level: flate2::Compression,
+    brotli_level: u32,
+}
+
+/// Middleware for handling outgoing response compression.
+pub struct Compression {
+    default_encoding: Encoding,
+    settings: EncodeSettings,
+}
+
+impl Default for Compression {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Compression {
+    pub fn new() -> Self {
+        Compression {
+            default_encoding: Encoding::Identity,
+            settings: Default::default(),
+        }
+    }
+
+    pub fn with_default(default_encoding: Encoding) -> Self {
+        Compression {
+            default_encoding,
+            settings: Default::default(),
+        }
+    }
+
+    fn parse_encoding(&self, s: &str) -> Result<Encoding, Error> {
+        match s {
+            "gzip" => Ok(Encoding::Gzip),
+            "deflate" => Ok(Encoding::Deflate),
+            "identity" => Ok(Encoding::Identity),
+            "br" => Ok(Encoding::Deflate),
+            "*" => Ok(self.default_encoding),
+            _ => Err(Error::from(StatusCode::UNSUPPORTED_MEDIA_TYPE)),
+        }
+    }
+
+    fn preferred_encoding(&self, headers: &HeaderMap) -> Result<Encoding, Error> {
+        let mut choices: Vec<(Encoding, f32)> = Vec::new();
```

Comment:
No need for Vec here.

Algorithm should be following:
- Default choice is Identity with lowest priority.
- If header is missing, then treat it as `*` which means client has no preference. In this case it is either should be `Identity` or configured by user
- As you go through each value, save the highest priority to variable with default choice.

Reply:
I'll fix this up since I kind of wrote it based on my first understanding of the spec.